proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1565 @@
+
+## Terminology / Concepts
+
+- **assertion** — a function that — given a value — if the value meets the assertion's criteria, quietly succeeds, but if the value fails the criteria, fails.
+- **CEL** — [Common Expression Language](https://github.com/google/cel-spec) — language supported (in alpha) by Kubernetes for CRD validation rules.
+- **module** — a file that can be loaded into a template in order to reuse its functions (or variables). The `ytt` "Standard Library" is comprised of modules that begin with the `@ytt:` prefix (e.g. @ytt:data, @ytt:assert, @ytt:overlay, ...) 
+- **node** — a single piece of YAML: a document set, document, map, map item, array, array item. One "annotates" a node by placing a `ytt` annotation-shaped comment just above the node (see also: [YAML Primer](https://carvel.dev/ytt/docs/v0.39.0/yaml-primer/)).
+- **required value** — a Data Value that the consumer must set. Mechanically, a Data Value that does not (yet) have a value that is "valid."
+- **rule** — the combination of: 
+    - a textual description of what constitutes a valid value;
+    - a function that asserts the rule against an actual value;
+    - a violation message template, used to report invalid values.
+    - **custom rule** — a rule based on a function (either one written by a user or provided in a `ytt` module).
+    - **named rule** — a rule provided with `ytt` expressed through a keyword argument to an `@schema/validation` or `@assert/validate`.
+- **schema base document** — the document instantiated from a schema; when the schema describes Data Values, it is the annotated default data values document.
+- **validation** — the binding of one or more "rule"s to a "node".
+- **violation** — an instance of where a given value of a "node" fails a "validation".
+
+## Proposal
+
+Implement all plumbing logic required to provide high-quality data validation over Data Values and output documents.
+
+Mechanically, this means:
+
+Extend `@schema/...` to be able to annotate validations in schema:
+- introduce `@schema/validation` which attaches a validation to a type.
+- when generating default Data Values (generally, the schema base document), when a type has a validation, annotate the corresponding node with an `@assert/validate` with that validation.
+
+Extend the `@ytt:assert` module, to include the ability to validate:
+- introduce an annotation (`@assert/validate`) which attaches one or more "rules" to a "node", defining what a valid value is for that node;
+- supply a set of common assertions (functions in the `@ytt:assert` module) that covers most validation use-cases;
+- provide guidance for how to express most commonly intended validations.
+
+Further, integrate this feature into `ytt`'s processing pipeline:
```

Comment:
Great question!  I can see where the proposal doesn't make this plain.

In practice, it will appear exactly as you said:
- in `@data/values-schema` documents, use the `@schema/validation` annotation.
- in all other documents, use the `@assert/validate` annotation.

In fact, an `@assert/validate` can be placed on _any_ node of _any_ kind of document and it would get evaluated.
Even deeper, we play a trick: turns out when we pull the "default data values" from the Schema?... well, we convert those `@schema/validation` into `@assert/validate` under the covers... so really, only the `@assert/validate` annotation is affecting any kind of validation.